documentation/advanced/tutorial-i18n-localization.asciidoc

```diff
@@ -169,3 +169,139 @@ public class MyLocale extends Div {
     }
 }
 ----
+
+=== Supporting Right-to-Left mode in the application
+
+Vaadin components have now support for RTL languages, so to make your
+application handle this mode, a few changes are needed:
+
+On top of the last examples, let's say that your application is now also
+translated into a RTL Language, such as Arabic. Besides <<Provider sample for
+translation, providing the sample for the Arabic translation>>, at your main
+layout you can add a code like such:
+
+[source,java]
+----
+public class MainLayout extends VerticalLayout {
+
+    public MainLayout() {
+        // ...
+        final UI ui = UI.getCurrent();
+        if (ui.getLocale().getLanguage() == "ar") {
+            ui.setDirection(Direction.RIGHT_TO_LEFT);
+        }
+    }
+}
+----
+
+This will work if changing of locale is only based on the `Accept-Language`
+coming from the client, but if user can define their language, for instance,
+on your application's setting page, then you can make your main layout implement
+the `LocaleChangeObserver` interface, so it will receive the changes of locale
+and then you can set the direction based on the locale set:
+
+[source,java]
+----
+public class MainLayout extends VerticalLayout implements LocaleChangeObserver {
+
+    @Override
+    public void localeChange(LocaleChangeEvent event) {
+        if (event.getLocale().getLanguage() == "ar") {
+            event.getUI().setDirection(Direction.RIGHT_TO_LEFT);
+        } else {
+            event.getUI().setDirection(Direction.LEFT_TO_RIGHT);
+        }
+    }
+}
+----
+
+==== Frontend projects
+
+For frontend applications only, to set the RTL mode you can call `document.dir = 'rtl'`.
+
+==== Adding RTL support to your custom elements
+
+If you have your own custom elements, there are a few steps needed in order to
+add RTL support to them:
+
+. If your element extends Vaadin's `ElementMixin`, make sure that it depends on
+the version starting from `2.3.0`. If not, you can make the element extends it
+or `DirMixin` only (`DirMixin` is part of the `@vaadin/vaadin-element-mixin`
+package).
+
++
+[source,javascript]
+----
+import { PolymerElement } from '@polymer/polymer/polymer-element.js';
+import { DirMixin } from '@vaadin/vaadin-element-mixin/vaadin-dir-mixin.js';
+
+class MyElement extends DirMixin(PolymerElement) {}
+----
+
++
+The `DirMixin` registers the element to changes on the `dir` attribute at the
+document level and keeps it in sync with the element's `dir` attribute. That's
+helpful to easily check the RTL status in both CSS and JS code.
+
+. Make sure your styles are adjusted properly for the RTL mode.
```

Comment:
I think we should mention logical properties (e.g. `margin-inline-start`), but add a disclaimer that those can only be used if you don’t have to support older browsers.

Reply:
Added more info about CSS Logical Properties with link.